### PR TITLE
Create /upload-artifacts route

### PR DESCRIPTION
**Pull request by prithvikannan**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Create `/upload-artifacts` route to update the eval table in the prompt engineering UI.

## How is this patch tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/tracking/test_rest_tracking.py [hunk 1]
```diff
@@ -1836,3 +1836,28 @@ def assert_response(resp, message_part):
...
+        f"{mlflow_client.tracking_uri}/ajax-api/2.0/mlflow/upload-artifact",
+        params={
+            "run_uuid": created_run.info.run_id,
+            "path": "test.txt",
+        },
+        data="hello world",
+    )
+    assert response.status_code == 200
+
+    response = requests.get(
```
Comment (harupy) [inline tests/tracking/test_rest_tracking.py:1855]:
```suggestion
    response = requests.post(
```

post?

File: mlflow/server/handlers.py [hunk 2]
```diff
@@ -1230,6 +1230,70 @@ def assert_arg_exists(arg_name, arg):
...
+        file_path = os.path.join(dir_path, basename)
+
+        # Ensure directory exists
+        if dirname:
+            os.makedirs(dir_path, exist_ok=True)
+
+        with open(file_path, "wb") as f:
+            f.write(data)
+            f.flush()
+            f.seek(0)
```
Comment (harupy) [inline mlflow/server/handlers.py:1289]: Curious why we need flush and seek. Can we pass `fiel_path` to `_log_artifact_to_repo`?

File: mlflow/server/handlers.py [hunk 3]
```diff
@@ -1230,6 +1230,70 @@ def assert_arg_exists(arg_name, arg):
...
+            error_code=INVALID_PARAMETER_VALUE,
+        )
+    path = args.get("path")
+    if not path:
+        raise MlflowException(
+            message="Request must specify path.",
+            error_code=INVALID_PARAMETER_VALUE,
+        )
+
+    if request.content_length > 10 * 1024 * 1024:
```
Comment (harupy) [inline mlflow/server/handlers.py:1250]: Any chance that `request.content_length` becomes None?

File: mlflow/server/handlers.py [hunk 4]
```diff
@@ -1230,6 +1230,70 @@ def assert_arg_exists(arg_name, arg):
     return response
 
 
+@catch_mlflow_exception
+def upload_artifact_handler():
+    _logger.info("upload_artifact_handler")
```
Comment (harupy) [inline mlflow/server/handlers.py:1235]: Can we remove unnecessary logging?

File: mlflow/server/handlers.py [hunk 5]
```diff
@@ -1230,6 +1230,70 @@ def assert_arg_exists(arg_name, arg):
...
+            path_to_log = dirname
+
+        artifact_repo.log_artifact(f.name, path_to_log)
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+        dir_path = os.path.join(tmpdir, dirname) if dirname else tmpdir
+        file_path = os.path.join(dir_path, basename)
+
+        # Ensure directory exists
+        if dirname:
```
Comment (harupy) [inline mlflow/server/handlers.py:1283]: Is this if necessary? Looks like `dirname` is always truthy.

File: tests/tracking/test_rest_tracking.py (hunk 1 above)
Comment (harupy) [reply tests/tracking/test_rest_tracking.py:1855]: never mind!

File: mlflow/server/handlers.py [hunk 6]
```diff
@@ -1230,6 +1230,70 @@ def assert_arg_exists(arg_name, arg):
...
+    _logger.info("upload_artifact_handler")
+    args = request.args
+    run_uuid = args.get("run_uuid")
+    if not run_uuid:
+        raise MlflowException(
+            message="Request must specify run_uuid.",
+            error_code=INVALID_PARAMETER_VALUE,
+        )
+    path = args.get("path")
+    if not path:
```
Comment (harupy) [inline mlflow/server/handlers.py:1244]: Is an empty string valid for `path`?

File: mlflow/server/handlers.py (hunk 3 above)
Comment (dbczumar) [reply mlflow/server/handlers.py:1250]: It probably can :)

**Review by dbczumar — APPROVED** [review]
LGTM once @harupy 's comments are addressed. Thanks @prithvikannan !

File: mlflow/server/handlers.py (hunk 6 above)
Comment (prithvikannan) [reply mlflow/server/handlers.py:1243]: No. Empty string for path should throw. I also call `validate_path_is_safe` to be safe.

File: mlflow/server/handlers.py [hunk 7]
```diff
@@ -1230,6 +1230,71 @@ def assert_arg_exists(arg_name, arg):
...
+    with tempfile.TemporaryDirectory() as tmpdir:
+        dir_path = os.path.join(tmpdir, dirname) if dirname else tmpdir
+        file_path = os.path.join(dir_path, basename)
+
+        os.makedirs(dir_path, exist_ok=True)
+
+        with open(file_path, "wb") as f:
+            f.write(data)
+
+        _log_artifact_to_repo(file_path, run, dirname, artifact_dir)
```
Comment (harupy) [inline mlflow/server/handlers.py:1292]: If `path` is like `foo/bar.txt`, `dirname` is `foo`. We create `foo` in the current working directory. Is this intended?

Comment (harupy) [reply mlflow/server/handlers.py:1292]: Is `foo/bar.txt` a valid path?

Comment (harupy) [reply mlflow/server/handlers.py:1292]: Ah, never mind.

**Review by harupy — APPROVED** [review]
LGTM!

**Pull request merged**